### PR TITLE
Google Analytics events for opening and closing the app, and opening and closing projects

### LINKED ISSUE #9778: Add Google Analytics for Open/Close App and Open/Close Workflow
- `open_app` should fire when the app is opened - both on the desktop app and the website.
- `close_app` should fire when the app is navigated away from.
- `open_workflow` should fire when the editor is opened.
- `close_workflow` should fire when the editor is closed, or the currently open project is changed.

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Closes #9778
  - Add `open_app`, `close_app`, `open_workflow`, and `close_workflow` events
- Miscellaneous fixes for Google Analytics:
  - Fix `open_chat` and `close_chat` events firing even when chat is not visible
  - Add Google Analytics script to GUI2 entrypoint (i.e. the entrypoint used by the desktop app)

Unrelated changes:
- Add Nix development shell to allow Nix users to build GUI2 and the build script
  - Java dependencies have *not* been added in this PR to keep things simple

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

File: app/gui2/index.html [hunk 1]
```diff
@@ -19,5 +19,7 @@
     <div id="enso-dashboard" class="enso-dashboard"></div>
     <div id="enso-chat" class="enso-chat"></div>
     <script type="module" src="/src/entrypoint.ts"></script>
+    <!-- Google tag (gtag.js) -->
+    <script async src="https://www.googletagmanager.com/gtag/js?id=G-CLTBJ37MDM"></script>
```
Comment (MrFlashAccount) [inline app/gui2/index.html]: Can we make it possible to not include gtm in dev mode?

File: app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx [hunk 2]
```diff
@@ -238,6 +316,25 @@ export default function AuthProvider(props: AuthProviderProps) {
     },
     [userSession?.type]
   )
+  const gtagEventRef = React.useRef(gtagEvent)
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx:244]: I've seen this snippet 2 times in the PR, can we extract this logic into a hook?

**Review by MrFlashAccount — APPROVED** [review]

File: app/gui2/index.html (hunk 1 above)
Comment (MrFlashAccount) [reply app/gui2/index.html]: Also, I think it makes sense to grab gtag ID from .env, because we might want to use different ids for different environments

Comment (somebody1234) [reply app/gui2/index.html]: re: moving it to .env - makes sense, but maybe after the environment variable is added to the GitHub runners -w which may mean it'll have to be done in another PR.

the dev mode thing can be done in this PR though.

File: app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/providers/AuthProvider.tsx:244]: oh i guess you mean the entire code block? yeah makes sense - next time you might want to shift-click the last line to comment on the entire block though

File: app/gui2/index.html (hunk 1 above)
Comment (somebody1234) [reply app/gui2/index.html]: on second thought: *for now, as a stopgap*, i think it's better to unconditionally include the tag id. when it gets added to the builders in GH Actions, then we can *omit* them in the `.env` files, and manually provide the environment variable when we want to test Google Analytics. thoughts?

**Comment by somebody1234** [comment]
pinging @MrFlashAccount for re-review.

File: app/gui2/index.html (hunk 1 above)
Comment (MrFlashAccount) [reply app/gui2/index.html]: Agree

**Review by MrFlashAccount — APPROVED** [review]

**Pull request merged**